pkg/kubelet/dockertools/labels.go

```diff
@@ -33,216 +34,229 @@ import (
 //  * label setters and getters
 //  * label filters (maybe in the future)
 
+// Version label indicates the version of labels on a container. Whenever we make incompatible changes
+// to the labels, we should bump up the current label version.
 const (
-	kubernetesPodNameLabel                   = "io.kubernetes.pod.name"
-	kubernetesPodNamespaceLabel              = "io.kubernetes.pod.namespace"
-	kubernetesPodUIDLabel                    = "io.kubernetes.pod.uid"
-	kubernetesPodDeletionGracePeriodLabel    = "io.kubernetes.pod.deletionGracePeriod"
-	kubernetesPodTerminationGracePeriodLabel = "io.kubernetes.pod.terminationGracePeriod"
-
-	kubernetesContainerNameLabel                   = "io.kubernetes.container.name"
-	kubernetesContainerHashLabel                   = "io.kubernetes.container.hash"
-	kubernetesContainerRestartCountLabel           = "io.kubernetes.container.restartCount"
-	kubernetesContainerTerminationMessagePathLabel = "io.kubernetes.container.terminationMessagePath"
-	kubernetesContainerPreStopHandlerLabel         = "io.kubernetes.container.preStopHandler"
+	versionLabel        = "io.kubernetes.label-version"
+	currentLabelVersion = "v1"
+)
 
-	// TODO(random-liu): Keep this for old containers, remove this when we drop support for v1.1.
-	kubernetesPodLabel = "io.kubernetes.pod.data"
+// Labels supported now
+const (
+	// Labels added from kubernetes v1.1
+	// Notice that the podNameLabel contains namespaced pod name in v1.1. However because we don't
+	// really rely on the pod name now, it should be fine.
+	// TODO(random-liu): Remove the comment above when we drop support for v1.1
+	podNameLabel       = "io.kubernetes.pod.name"
+	containerNameLabel = "io.kubernetes.container.name"
 
+	// Labels added from kubernetes v1.2
+	podNamespaceLabel              = "io.kubernetes.pod.namespace"
+	podUIDLabel                    = "io.kubernetes.pod.uid"
+	containerHashLabel             = "io.kubernetes.container.hash"
 	cadvisorPrometheusMetricsLabel = "io.cadvisor.metric.prometheus"
+
+	// Labels added from kubernetes v1.3
+	containerInfoLabel = "io.kubernetes.container.info"
 )
 
-// Container information which has been labelled on each docker container
-// TODO(random-liu): The type of Hash should be compliance with kubelet container status.
-type labelledContainerInfo struct {
-	PodName                   string
-	PodNamespace              string
-	PodUID                    types.UID
-	PodDeletionGracePeriod    *int64
-	PodTerminationGracePeriod *int64
-	Name                      string
-	Hash                      string
-	RestartCount              int
-	TerminationMessagePath    string
-	PreStopHandler            *api.Handler
+// containerMeta is the metadata we need to labelled on each docker container. Each field will be one
+// label on the container. We could also use the label to select specific containers in the future.
+// Notice that the container meta should seldom or never change.
+type containerMeta struct {
```

Comment:
@vishh I found that @sjpotter is working on move the "metadata" labels to kubelet level #25688 for summary api consuming.

If we still want to group these values into one label, we could:
* Group these labels to a type called `ContainerMeta` and move the type to `container/` directory
* Add function in `container/` to deserialize `ContainerMeta`, and make sure all runtime serializes it with the same format (maybe json) when creating containers.
* Change #25688 to use the new functions.
* Add backward compatible logic in docker labels.

This is doable, but it looks like it doesn't worth the effort. The only concern of keeping them in separate labels is version problem. However, we won't change the metadata too often (in fact almost never change), so there shouldn't be version issue even if we keep them in separate labels.

Moreover, keeping them in different labels makes it possible to do flexible selecting when we have other requirements in the future, and won't break user who really uses these labels #14768.